### PR TITLE
switch block added

### PR DESCRIPTION
**Pull request by asafbennatan**
related issue #181 
allows the following template:

```
@args (String s)
@switch (s) {
case("test"){ this is a test}
case("test2"){ this is a test2}
default{ this is default }
}
```
or

```
@args (String s)
@switch (s) {
case("test"){ this is a test}
case("test2"){ this is a test2}
   }
```

this will generate a java switch case code:

```
switch(s){
  case "test" ->{ ...}
  case "test2" -> {...}
  default ->{....}
}
```

and : 

```
switch(s){
  case "test" ->{ ...}
  case "test2" -> {...}
}
```

### CODE REVIEW

**Comment by jjlauer** [comment]
This is pretty awesome! I will try to give it a deep dive in the next few days and get you some feedback and/or merge.

**Comment by asafbennatan** [comment]
@jjlauer thanks, LMK if you have any update/ questions etc

**Comment by jjlauer** [comment]
Have you run the unit tests with your changes? I current get a failure in one of the unit tests with your branch.

**Comment by asafbennatan** [comment]
hi , i am running on windows so not sure if this is related but running the latest commit on main repo's master results in a failing test:

```
 Tests run: 37, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 0.525 s <<< FAILURE! - in com.fizzed.rocker.compiler.TemplateParserTest
[ERROR] com.fizzed.rocker.compiler.TemplateParserTest.testIfElseStatement  Time elapsed: 0.015 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is "}\nelse {\nspacy\n}"
     but: was "}\r\nelse {\r\nspacy\r\n}"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:964)
	at org.junit.Assert.assertThat(Assert.java:930)
	at com.fizzed.rocker.compiler.TemplateParserTest.testIfElseStatement(TemplateParserTest.java:790)
```

this is from commit 83499963

is that the one failing for you ? is this a windows issue ?

**Comment by jjlauer** [comment]
Yeah, those two unit tests were failing on just **some** versions of windows.  I just pushed up a fix for them that worked on a windows machine that I was able to reproduce them on.  If you pull in those changes to your branch, you should now see the 1 unit test failing with your branch.

**Comment by asafbennatan** [comment]
yes now i get test pass from repo master branch and i see the failing test on my branch , it is this one:

```
[ERROR] Tests run: 39, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.598 s <<< FAILURE! - in com.fizzed.rocker.compiler.TemplateParserTest
[ERROR] com.fizzed.rocker.compiler.TemplateParserTest.plainTextIncludesStyleWithNoMatchingRightCurly  Time elapsed: 0.008 s  <<< FAILURE!
java.lang.AssertionError: expected exception
	at org.junit.Assert.fail(Assert.java:89)
	at com.fizzed.rocker.compiler.TemplateParserTest.plainTextIncludesStyleWithNoMatchingRightCurly(TemplateParserTest.java:307)

```

**Comment by asafbennatan** [comment]
@jjlauer fixed

File: rocker-compiler/src/test/resources/rocker/parser/SwitchBlockWithDefault.rocker.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
+
+@args (String s)
+@switch (s) {
+case("test")   { this is a test}
```
Comment (jjlauer) [inline rocker-compiler/src/test/resources/rocker/parser/SwitchBlockWithDefault.rocker.html:4]: Does the ( parenthese next to the case keyword need to be directly next to it to work? In most of the rocker docs, we generally allow whitepsace in similar situations for if/else blocks.  Same feedback on the "default", whitespace before the next { bracket.

So I'd suggest you add just a couple more use cases in these examples with whitespace for "case ("val")" and maybe another unit test demonstrating "default { }" with whitespace also works.

Also, some examples of using keywords IN the html blocks would be good to verify as well.  So inside a "case" block, output the word "case" as html to make sure the parser doesn't break.  I'd also throw the keyword "default" into the html blocks to validate those don't break either.

**Review by jjlauer — COMMENTED** [review]
All unit tests pass and I also ran your branch against a few very large codebases that use rocker and those all compiled as well.  So that's a pretty solid validation these changes are solid.

I only had some feedback on a few additional unit tests I think are worth adding mostly around validating whitespace in some cases, and double checking keywords like case/default/switch don't confuse the parser inside the html blocks.

Awesome job though so far! Very excited to get this shipped.

File: rocker-compiler/src/test/resources/rocker/parser/SwitchBlockWithDefault.rocker.html (hunk 1 above)
Comment (asafbennatan) [reply rocker-compiler/src/test/resources/rocker/parser/SwitchBlockWithDefault.rocker.html:4]: will add more test cases , as for whitespaces it should be fine (but ill verify).
regarding whitespace inside the switch this caused an issue (since it will generate invalid code) the way i solved this was by removing any whitespace (and failing for non whitespace) at switch exit:

see  exitSwitchBlock calling verifySwitchBlock

Comment (asafbennatan) [reply rocker-compiler/src/test/resources/rocker/parser/SwitchBlockWithDefault.rocker.html:4]: @jjlauer done , LMK if this is what you meant ( see rocker/parser/SwitchBlockWithDefaultSpaces.rocker.html and  testSwitchBlockWithDefaultSpaces)

**Review by jjlauer — APPROVED** [review]

**Comment by jjlauer** [comment]
Unit tests look great. Merged your PR.  Will publish a new release hopefully sometime today.
